### PR TITLE
Adding support for arbritrary NBT tags

### PR DESCRIPTION
**Pull request by 12xx12**
see #4179

Added Compound and types
Added Conversion for cParsedNBT and cFastNBT

### CODE REVIEW

**Comment by 12xx12** [comment]
AAAAAAAAAA - this compiles in MVSC :/

File: src/WorldStorage/NBT.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+class cFastNBTWriter;
+
+// fwd:
+class cNBTContent;
+class cNBT;
+namespace NBT
+{
+	using cByteArray = std::vector<char>;  // Todo: Make this a Int8
+	using cIntArray = std::vector<Int32>;
+	using cList = std::vector<cNBT>;
```
Comment (peterbell10) [inline src/WorldStorage/NBT.h:22]: A list isn't just an array of tags, all the entries need to have the same tag type.

Comment (tigerw) [reply src/WorldStorage/NBT.h:22]: cNBT::Push enforces this I believe

Comment (12xx12) [reply src/WorldStorage/NBT.h:22]: Actually yes, BUT you can create a cList object and create a cNBT from this and have a list with various types of entries

**Comment by 12xx12** [comment]
Hm. Clang doesn't like the circular behaviour.

I'd suggest creating a cCompound class that stores dynamically allocated cNBT objects in a map. Directly using a unordered map works too but then the user has to keep track of the memory leaks. I'd rather use the cCompound to take care of this

**Comment by tigerw** [comment]
What's the issue with circular things? We shouldn't need any sort of inheritance right?

**Comment by 12xx12** [comment]
We habe compounds including cNBT objects

And we have cNBT objects containing the variant containing compounds

**Comment by tigerw** [comment]
It should just work, I think. I.e. with the example in the other PR this compiles:

```cpp
    NBT Root{
        NBT::Compound{
            { "a", { NBT::List() } },
            { "b", { 5 } }
        }
    };
```

**Comment by tigerw** [comment]
Of course it depends on undefined behaviour of std::map accepting incomplete types; so we should probably look to storing a std::unique_ptr in an unordered_map for your solution

**Comment by 12xx12** [comment]
ok - i made the variant contain a pointer to ```cCompund```.  Now the default constructor of ```cNBT``` creates a ```cCompund``` with new and deletes it

**Comment by 12xx12** [comment]
I think i'm gonna hide some things in the private parts (yeah sorry) of the ```cNBT``` class

File: src/WorldStorage/NBT.h (hunk 1 above)
Comment (12xx12) [reply src/WorldStorage/NBT.h:21]: i tried to fix this

File: src/WorldStorage/FastNBT.h [hunk 2]
```diff
@@ -21,31 +21,8 @@ It directly outputs a string containing the serialized NBT data.
...
-	TAG_List      = 9,
-	TAG_Compound  = 10,
-	TAG_IntArray  = 11,
-	TAG_Max       = 11,  // The maximum value for a tag type
-} ;
-
-
-
+#include "NBTDef.h"
+#include "NBT.h"
```
Comment (12xx12) [inline src/WorldStorage/FastNBT.h:25]: remove this

**Comment by 12xx12** [comment]
wooooooooooooooooop yeeeeeaaaaaaaaaaaaahhhhhhhhhh

thanks @tigerw

File: src/WorldStorage/NBT.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	using cNBTContent = std::variant<
+		cEmptyTag,      // TAG_End
+		unsigned char,  // TAG_Byte  // Todo: Make this a Int8
+		Int16,          // TAG_Short
+		Int32,          // TAG_Int
+		Int64,          // TAG_Long
+		float,          // TAG_Float
+		double,         // TAG_Double
+		AString,        // TAG_String
+		cCompound *,    // TAG_Compound
```
Comment (12xx12) [inline src/WorldStorage/NBT.h:35]: Maybe a pointer to `cNBT` would be better to assure safty even for nested lists

Comment (12xx12) [reply src/WorldStorage/NBT.h:35]: I'm trying to use a smart pointer but that blows up :/

**Comment by tigerw** [comment]
There was discussion on your forum thread (https://forum.cuberite.org/thread-3325.html), the main outcome was storing a std::variant containing every thing we need (as objects) into cItem, for better type safety and more ergonomic field accesses. If we do that a cNBT class isn't immediately required anymore.

**Comment by 12xx12** [comment]
But do we access the values in the vector of variants - do we set constant indexes for certain values? Or do we actually map values from string to values?

**Comment by tigerw** [comment]
We can say `std::get<PropertyClassHere>(Item.Properties)`

**Comment by 12xx12** [comment]
Didn't know this works on a vector of variant? What is the behaviour if there are multiple values of the same type?

**Comment by tigerw** [comment]
Oh I see what you mean, we do a linear search and the first variant that matches counts. A map/set is conceptually more appropriate but I'd say is too high overhead for a couple largely fixed values.

**Comment by 12xx12** [comment]
Yeah this makes more sense - there won't be that many entries in the vector so that linear search wont be that slow

**Comment by 12xx12** [comment]
Closed in favour of #5454

**Pull request closed without merging**